### DISCUSSION TITLE
How to use latex's `csquotes`?

### DISCUSSION THREAD

**Discussion by putzwasser** [quarto-dev/quarto-cli#11673] [Q&A]
### Description

Hi there,

I'm starting to write an academic paper using quarto. I'm stuck at how to properly create quotes. I.e., how can I use the `csquotes` package?

I created this minimal document:

```md
---
title: "Your Document"
format:
  pdf:
    keep-tex: true
    header-includes: \usepackage{csquotes}
---

Here is an inline quote: <q>This is quoted text</q>.

> This is a block quote.

```

Running quarto didn't yield the result I hoped for:

```bash
quarto render test.qmd -t pdf --log-level debug
Quarto version: 1.6.39

pandoc --verbose --trace
  to: latex
  output-file: test.tex
  standalone: true
  pdf-engine: xelatex
  variables:
    graphics: true
    tables: true
  default-image-extension: pdf
  
metadata
  documentclass: scrartcl
  classoption:
    - DIV=11
    - numbers=noendperiod
  papersize: letter
  header-includes:
    - \KOMAoption{captions}{tableheading}
    - \usepackage{csquotes}
  block-headings: true
  title: Your Document

# ...

[INFO] Not rendering RawInline (Format "html") "<q>"
[INFO] Not rendering RawInline (Format "html") "</q>"

# ...

```

This render run gave me the following latex source code which is was used to create the PDF.

```latex
%...

\begin{document}
\maketitle


Here is an inline quote: This is quoted text.

\begin{quote}
This is a block quote.
\end{quote}

\end{document}
```

This makes me wonder:

- Why does quarto ignore the HTML inline quote `q` tags?
- Is there another way or syntax to create inline quotes using markdown and quarto? (I just found [block quotes](https://quarto.org/docs/authoring/markdown-basics.html#other-blocks) and [citations](https://quarto.org/docs/authoring/citations.html) in the docs)
- (How) can I tell quarto that block quotes should go into a `displayquote` environment instead of a regular `quote` environment?

Thank you in advance for taking the time to read an answer.

**Comment by cscheid** [comment]
> Why does quarto ignore the HTML inline quote q tags?

That's not Markdown. Why do you expect `<q>` to work in non-html formats?

**Reply by putzwasser** [reply]
My thinking was:

1. Markdown doesn't provide a shortcode for inline quotes.
2. Markdown supports raw HTML tags.
3. Quarto states it supports scientific markdown at the home page.

This is why I (naively) thought/assumed using `<q>` to create inline quotes for academic writing would be supported.

**Reply by mcanouil** [reply]
> Markdown supports raw HTML tags.

Where did you see that?
That's not true.

**Reply by putzwasser** [reply]
> > Markdown supports raw HTML tags.
> 
> Where did you see that?

https://daringfireball.net/projects/markdown/syntax#html

> For any markup that is not covered by Markdown’s syntax, you simply use HTML itself. There’s no need to preface it or delimit it to indicate that you’re switching from Markdown to HTML; you just use the tags.

AFAIK, most markdown implementations support inserting HTML. (see [CommonMark](https://spec.commonmark.org/0.31.2/#raw-html) spec)

But I'm not sure how discussing this helps here.

I had wrong assumptions about how quarto handles my markdown. How can I configure it, so that `q` tags are supported properly?

**Reply by mcanouil** [reply]
If you read that documentation, it's "text to html" thus the fact that raw html can work in this context.
You are not in this context here.
You are in "text to latex".

**Reply by mcanouil** [reply]
More generally, you are in the context of Pandoc/Quarto which is "text to many formats", so the correct/accurate documentation are:
- https://pandoc.org/MANUAL.html#pandocs-markdown
- https://quarto.org/docs/authoring/markdown-basics.html

**Reply by putzwasser** [reply]
> You are in "text to latex".

Okay, I get that distinction.

https://quarto.org/docs/authoring/markdown-basics.html doesn't say anything about inline quotes. I read it before coming here. Honestly, I didn't think that inline quotes weren't supported for creating PDFs by default somehow.

So, I tried an approach that I hoped might work. It didn't. Thus, I came here seeking help.

**Reply by mcanouil** [reply]
If the documentation (Pandoc and Quarto) does not say anything about inline quotes, it's simply because it does not exist.

You need to use raw LaTeX as described in the markdown basics guide I provided earlier.

**Reply by putzwasser** [reply]
And can I implement/add something so that `q` tags are supported?

I'm asking because

1. Some markdown live editors (at leas the one I use) render `q` tags and this makes authoring visually easier.
2. Using `q` tags would easily allow md → html, if I wanted to add an HTML rendering of the document at some point. If I was using raw LaTeX this would demand additional work.

Besides the `q` tag: Is there a way tell quarto that block quotes should go into a `displayquote` environment instead of a regular `quote` environment?

**Reply by mcanouil** [reply]
You could create a LUA filter to do the conversion for LaTeX and other non-HTML formats.

**Comment by cderv** [comment]
Pandoc has support for `csquotes: true` to use `\enquote{}` for any inline quoted text
````markdown
---
title: "Your Document"
format:
  pdf:
    keep-tex: true
csquotes: true
---

Here is an inline quote: "This is quoted text"

> This is a block quote.
````

It will use package `csquotes`
https://github.com/quarto-dev/quarto-cli/blob/83a3b5af86d5f43f5f611009b2af8e74e94ffad3/src/resources/formats/pdf/pandoc/latex.template#L479-L481

and result in 
````latex
Here is an inline quote: \enquote{This is quoted text}
````

Though it seems they don't replace `\begin{quote}` by `\begin{displayquote}`: It would be a feature request for Pandoc. 

> Is there a way tell quarto that block quotes should go into a displayquote environment instead of a regular quote environment?

The Lua filter route for a Quarto extension handling of csquote would be the way, by adding support for `\enquote` and handling of `BlockQuote` to insert TeX Raw Block using `displayquote`. 

Note that there is already https://github.com/quarto-ext/latex-environment that give generic support for LaTeX command and environment. 
````markdown
---
title: "Your Document"
format:
  pdf:
    keep-tex: true
    include-in-header: 
      text: \usepackage{csquotes}
filters:
   - latex-environment
environments: [displayquote]
commands: [enquote]
---

Here is an inline quote: [This is quoted text]{.enquote}

::: {.displayquote}
This is a block quote.
:::

````

gives 
````latex
Here is an inline quote: \enquote{This is quoted text}

\begin{displayquote}
This is a block quote.

\end{displayquote}
````

**Reply by putzwasser** [reply]
Thanks for your reply. This helps a lot.

As for inline quotes:

Using ChatGPT, I created this simple LUA filter:

```lua
function RawInline(el)
  -- Check if the element is an HTML inline element <q>
  if el.format == "html" then
    if el.text:match("^<q>$") then
      return pandoc.RawInline("latex", "\\textquote{")
    end
    if el.text:match("^</q>$") then
      return pandoc.RawInline("latex", "}")
    end
  end
end
```

It seems to work, but I guess this will be highly fragile. I'd welcome any ideas for improvement. Maybe this filter can make it into quartos core afterward (especially, in combination with the `csquotes: true` setting)? I guess, more people would benefit from this for academic writing.

> Though it seems they don't replace \begin{quote} by \begin{displayquote}: It would be a feature request for Pandoc.

[Created one](https://github.com/jgm/pandoc/issues/10456).

I also tried to create a Lua filter to handle blockquotes:

```lua
-- Lua filter to convert block quotes into \displayquote for LaTeX
function BlockQuote(el)
  -- Check if the target format is LaTeX
  if FORMAT == "latex" then
    -- Convert the block quote's content back into LaTeX
    local content = pandoc.write(el, "latex")
    -- Wrap the content in \displayquote environment
    return pandoc.RawBlock("latex", "\\begin{displayquote}\n" .. content .. "\n\\end{displayquote}")
  else
    -- Return the block quote unchanged for other formats
    return el
  end
end
```

But this throws an error:

```bash
Pandoc expected, got Block
	while retrieving Pandoc
	while retrieving function argument doc
	while retrieving arguments for function write
stack traceback:
	filters/replace-latex-quote-by-displayquote.lua:6: in function <filters/replace-latex-quote-by-displayquote.lua:2>
	[C]: in ?
	[C]: in method 'walk'
	/opt/quarto/share/filters/main.lua:558: in local 'checked_walk'
	/opt/quarto/share/filters/main.lua:610: in function 'run_emulated_filter'
	/opt/quarto/share/filters/main.lua:1297: in local 'callback'
	[string "..."]:1888: in field 'withScriptFile'
	/opt/quarto/share/filters/main.lua:1313: in upvalue 'run_emulated_filter_chain'
	/opt/quarto/share/filters/main.lua:1351: in function </opt/quarto/share/filters/main.lua:1348>
stack traceback:
	/opt/quarto/share/filters/main.lua:558: in local 'checked_walk'
	/opt/quarto/share/filters/main.lua:610: in function 'run_emulated_filter'
	/opt/quarto/share/filters/main.lua:1297: in local 'callback'
	[string "..."]:1888: in field 'withScriptFile'
	/opt/quarto/share/filters/main.lua:1313: in upvalue 'run_emulated_filter_chain'
	/opt/quarto/share/filters/main.lua:1351: in function </opt/quarto/share/filters/main.lua:1348>
Quarto render failed.
```

Using `local content = pandoc.utils.stringify(el)` would work, but it strips all formatting.

**Reply by cderv** [reply]
> Using ChatGPT, I created this simple LUA filter

Beware of doing that, unless you took the time to read Lua documentation for Quarto and Pandoc first. 
- https://quarto.org/docs/extensions/lua.html
- https://quarto.org/docs/extensions/lua-api.html

This quite specific with extended syntax compated to base LUA, and ChatGPT doesn't know that. 

> I also tried to create a Lua filter to handle blockquotes:

You could also just add the latex content as RawInline before and after the Blockquote content. I don't know what is the issue exactly, you need to try debug it (`quarto.log.output()` can help you test the different results you have. 

As hint I give you the following

> Pandoc expected, got Block

This error should be clear to you what is wrong, unless you did not read enough about Pandoc's Lua filter. 
See the doc of [`pandoc.write()`](https://pandoc.org/lua-filters.html#pandoc.write) that you are using, and look at what is the expected input. 

Hope it helps
